### ISSUE TITLE
fatal: Not a git repository

### ISSUE THREAD

**Issue by jozsi** [typicode/husky#68]
Hello,

I am trying to install husky inside a docker container that has a host directory mounted as a data volume.

Apparently, the working directory is set to `./node_modules/husky` instead of `.` so `hooksDir` fails:

```
root@f5412bf9a01f:/usr/src/web# npm install husky --save-dev
-
> husky@0.11.8 install /usr/src/web/node_modules/husky
> node ./bin/install.js

husky
setting up hooks in .git/hooks
  fatal: Not a git repository (or any parent up to mount point /usr/src/web/node_modules)
Stopping at filesystem boundary (GIT_DISCOVERY_ACROSS_FILESYSTEM not set).

husky@0.11.8 node_modules/husky
├── normalize-path@1.0.0
└── is-ci@1.0.9
root@f5412bf9a01f:/usr/src/web# pwd
/usr/src/web
root@f5412bf9a01f:/usr/src/web# git rev-parse --git-dir
.git
root@f5412bf9a01f:/usr/src/web# cd node_modules/husky/
root@f5412bf9a01f:/usr/src/web/node_modules/husky# git rev-parse --git-dir
fatal: Not a git repository (or any parent up to mount point /usr/src/web/node_modules)
Stopping at filesystem boundary (GIT_DISCOVERY_ACROSS_FILESYSTEM not set).
```

I need to manually run the following to get it to work:

```
root@f5412bf9a01f:/usr/src/web# node node_modules/husky/bin/install.js
husky
setting up hooks in .git/hooks
done
```

I suppose it should work on npm install as well.

PS: My npm version is `2.15.9`

**Comment by iamstarkov** [comment]
why do you need `husky` if you are not using `git`?

**Comment by piperchester** [comment]
@jozsi is this issue still relevant?

**Comment by jozsi** [comment]
@piperchester - I moved away from developing inside Docker. The key error is:
```
fatal: Not a git repository (or any parent up to mount point /usr/src/web/node_modules)
Stopping at filesystem boundary (GIT_DISCOVERY_ACROSS_FILESYSTEM not set).
```

It should traverse to `/usr/src/web/` which is the git repo and contains the source code.

The `node_modules` folder is mounted from a different volume than it's parent (source code). It's strange that running through `npm install` it hit that technical limitation, but it works correctly if I manually invoke the install script. I am unsure which one is to blame: Docker, npm, git or husky - or if it's still the issue by now. So I'll close it.

**Comment by szimek** [comment]
We had a similar problem. We're running our tests in a Docker image and to get advantage of Docker build cache, we first copy `yarn.lock` file and all `package.json` files (we're using lerna), then we install dependencies and only after all dependencies are install, we copy all remaining files, including git repo. This causes an issue with `husky`, because at the time it's installed, there's no git repo available yet. For now I fixed it by moving `husky` to `optionalDependencies` - this way the build doesn't break if installing `husky` fails.

**Comment by deltakroneker** [comment]
> We had a similar problem. We're running our tests in a Docker image and to get advantage of Docker build cache, we first copy `yarn.lock` file and all `package.json` files (we're using lerna), then we install dependencies and only after all dependencies are installed, we copy all remaining files, including git repo. This causes an issue with `husky`, because at the time it's installed, there's no git repo available yet. For now I fixed it by moving `husky` to `optionalDependencies` - this way the build doesn't break if installing `husky` fails.

Have you figured out the correct way to fix this?

**Comment by namcusamlc** [comment]
@szimek @deltakroneker I faced the same problem with installing `husky` inside `node docker` too.
The problem occurred when I moved `from husky v4 to v7`
My solution was to move husky install (`npx husky install`) to start command of the docker, also set `prepare` inside `scripts` to empty string `""`

Here are steps:
- add `npx husky install` to `command` of `docker-compose` or inside `CMD` of `Dockerfile`
- remove prepare script so that it won't throw fatal error every time we ran `npm install`, use this command: `npm set-script prepare ""`

Now everything works fine and all husky hooks are committed to `git`, my teammates and I have the same version of husky configs now.

Hope this will help people struggling the same problem.
